In [40]:
import requests
from bs4 import BeautifulSoup
from selenium import webdriver
import time
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from urllib.parse import urljoin
import numpy as np

### First Attempt just using Beautiful Soup

In [ ]:
levels = ['college', 'high-school']
sports = ['football', 'basketball', 'baseball', 'womens-basketball', 'volleyball', 'gymnastics',
         'mens-lacrosse', 'womens-lacrosse', 'mens-soccer', 'womens-soccer', 'softball', 'womens-track',
         'mens-golf', 'womens-golf', 'mens-hockey', 'womens-hockey', 'mens-swimming', 'womens-swimming']

all_names = []

for level in levels:
    for sport in sports:

        URL = f'https://www.on3.com/nil/rankings/player/{level}/{sport}/'

        page = requests.get(URL)
        soup = BeautifulSoup(page.content, 'html.parser')
        results = soup.find(class_="NilRankingsPageComponent_nilRankingsList__t14Ao")

        try:
            names_sport = results.find_all('a', class_="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-h5 MuiTypography-colorPrimary")
            for name in names_sport:
                all_names.append(name)
        except:
            # print("Whoops! No athletes appear to have NIL deal for {} {}".format(level, sport))
            continue
    #end
#end

In [7]:
print(len(all_names))

414


It looks like there were only 414 athletes using this section of the On3 database... This is such a small number of athletes that we should investiage looking at different sections of the website

### Using Selenium & Beatuiful Soup

In [65]:
sports = ['football']
#years = ['2022','2023','2024','2025','2026']
years = ['2023']

exp = []
athlete_grade = []
ages = []
ranks = []
high_school = []
home_town = []
colleges = []
college_status = []
college_distance = []
num_offers = []
NIL_val = []
instagram_followers = []
twitter_followers = []
tiktok_followers = []

for year in years:
    for sport in sports:
        
        # Selenium Driver to click on "Load More"
        driver = webdriver.Chrome()
        URL = f'https://www.on3.com/db/rankings/industry-player/{sport}/{year}/'
        driver.get(URL)
        
        # while True:
        #     try:
        #         load_more_button = WebDriverWait(driver, 10).until(
        #             EC.element_to_be_clickable((By.XPATH, "//span[@class='MuiButton-label' and contains(text(), 'Load More')]"))
        #         )
        #         load_more_button.click()
        #         time.sleep(10)
        #     except:
        #         break
        
        # Beautiful Soup - for each player
        page_source = driver.page_source
        page_soup = BeautifulSoup(page_source, 'html.parser')
        results = page_soup.find(class_="PlayerRankings_playerRankings__jvfFg")

        athletes = results.find_all('a', class_="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-h5 MuiTypography-colorPrimary")
        # It is possible that two athletes have the same name: John Smith
        # We cannot use a dictionary so a list of tuples is used for now
        names = [athlete.text for athlete in athletes]

        # Generate a list of links that we can iterate through
        links = [athlete['href'] for athlete in athletes]
        base_url = "https://www.on3.com/"
        athlete_links = [urljoin(base_url, link) for link in links]

        for athlete_link in athlete_links:
            # We can now go into each individual athletes page
            # for that sport, for that year, after loading more
            try:
                driver.get(athlete_link)
                time.sleep(3)

                athlete_source = driver.page_source
                page_soup = BeautifulSoup(athlete_source, 'html.parser')

                CollegeRankingInfo = page_soup.find(class_='CollegeRanking_info__LM3nn')

                if CollegeRankingInfo:
                    exp_year = CollegeRankingInfo.find_all(class_='MuiTypography-root CollegeRanking_span__qtAfW MuiTypography-subtitle1 MuiTypography-colorTextPrimary')
                    if len(exp_year) == 2:
                        exp.append(exp_year[0].text)
                        athlete_grade.append(exp_year[1].text)
                    else:
                        exp.append(np.nan)
                        athlete_grade.append(np.nan)
                    age = CollegeRankingInfo.find(class_='MuiTypography-root CollegeRanking_span__qtAfW MuiTypography-subtitle1 MuiTypography-colorPrimary')
                    ages.append(age.text)
                else:
                    exp.append(np.nan)
                    athlete_grade.append(np.nan)
                    ages.append(np.nan)
                
                RecruitModuleInfo = page_soup.find(class_='RecruitModule_info__Ugxqd')

                if RecruitModuleInfo:
                    ranking = RecruitModuleInfo.find(class_='RecruitModule_rating__sONqb')
                    ranks.append(ranking.text)

                    homeInfo = RecruitModuleInfo.find_all(class_='MuiTypography-root RecruitModule_span__KmmzN MuiTypography-subtitle1 MuiTypography-colorTextPrimary')
                    high_school.append(homeInfo[-2].text)
                    home_town.append(homeInfo[-1].text)
                else:
                    ranks.append(np.nan)
                    high_school.append(np.nan)
                    home_town.append(np.nan)
            except:
                print("Link {} is no good - removing athlete altogether".format(athlete_link))
                badLink = athlete_link
                index = athlete_link.index(badLink)
                names.pop(index)
                continue
            
            
            # Move to Recruiting Page
            driver.get(urljoin(athlete_link,'recruiting/'))
            time.sleep(3)
            athlete_source = driver.page_source
            page_soup = BeautifulSoup(athlete_source, 'html.parser')
            
            # Find url for all teams and navigate to page
            all_team_link = page_soup.find(class_='MuiTypography-root MuiLink-root MuiLink-underlineHover PlayerInterestsModule_text__kjqNU MuiTypography-caption MuiTypography-colorPrimary')
            driver.get(urljoin(base_url,all_team_link['href']))
            time.sleep(3)
            athlete_source = driver.page_source
            page_soup = BeautifulSoup(athlete_source, 'html.parser')
            
            try:
                RecruitColleges = page_soup.find_all(class_='PlayerInterestsItem_teamContainer__vjQkf')
                if RecruitColleges:
                    dummy_colleges=[]
                    count = 0
                
                    for college in RecruitColleges:
                        college_name = college.find(class_='MuiTypography-root MuiLink-root MuiLink-underlineNone PlayerInterestsItem_teamName__FeBHv MuiTypography-h5 MuiTypography-colorPrimary')
                        college_statuses = college.find(class_='MuiTypography-root PlayerInterestsItem_status__1_rA8 MuiTypography-subtitle1 MuiTypography-colorTextPrimary')
                        college_dist = college.find(class_='MuiTypography-root PlayerInterestsItem_distanceText__KJhj3 MuiTypography-caption MuiTypography-colorTextPrimary')
                        if count == 0:
                            colleges.append(college_name.text)
                            college_status.append(college_statuses.text)
                            college_distance.append(college_dist.text)
                        count +=1
                    num_offers.append(count)
                else:
                    colleges.append(np.nan)
                    college_status.append(np.nan)
                    college_distance.append(np.nan)
                    num_offers.append(np.nan)
                    
            except:
                print("error on recruiting page")
            
            # Move to NIL Page
            driver.get(urljoin(athlete_link,'nil/'))
            time.sleep(3)
            athlete_source = driver.page_source
            page_soup = BeautifulSoup(athlete_source, 'html.parser')
            
            try:
                RecruitNIL = page_soup.find(class_='NilValuationCircle_nilCircleValue__wzomB')
                if RecruitNIL:
                    NIL_val.append(RecruitNIL.text)
                else:
                    NIL_val.append(np.nan)
                    
            except:
                print('error on NIL val')
            
            try:       
                #Need to figure out a way to get each social that is available and determine which it is from the URL possibly
                RecruitSocials = page_soup.find(class_="NilSocialValuations_socialValuations__MeR7O")
                if RecruitSocials:
                    socials = RecruitSocials.find_all(class_="NilSocialValuations_platform__3qBMy")
                    instagram = np.nan
                    twitter = np.nan
                    tiktok =np.nan
                    for social in socials:
                        social_name = social.find('a')['href']
                        if 'instagram' in social_name:
                            instagram = social.find(class_='MuiTypography-root NilSocialValuations_platformFollowers__sW8kK MuiTypography-body1 MuiTypography-colorTextPrimary').text
                        elif 'twitter' in social_name:
                            twitter = social.find(class_='MuiTypography-root NilSocialValuations_platformFollowers__sW8kK MuiTypography-body1 MuiTypography-colorTextPrimary').text
                        elif 'tiktok' in social_name:
                            tiktok = social.find(class_='MuiTypography-root NilSocialValuations_platformFollowers__sW8kK MuiTypography-body1 MuiTypography-colorTextPrimary').text
                    
                else:
                    instagram = np.nan
                    twitter = np.nan
                    tiktok =np.nan
                
                instagram_followers.append(instagram)
                twitter_followers.append(twitter)
                tiktok_followers.append(tiktok)
                
            except:
                print("error on social media")
                
        #end
        print(len(names), len(exp), len(athlete_grade), len(ages), len(ranks), len(high_school), len(home_town))
        print(names[0], exp[0], athlete_grade[0], ages[0], ranks[0], high_school[0], home_town[0])
    #end
#end

        # This successfully takes us back to the previous page, but it might require us to reload every athlete...
        # driver.back()
        # url = driver.current_url
        # print(url)

driver.quit()

50 50 50 50 50 50 50
Arch Manning 2023 - present Freshman - 99.53 Isidore Newman New Orleans, LA


In [69]:
print(instagram_followers)
print(twitter_followers)
print(tiktok_followers)

['200K', '15.4K', '42K', '136K', '82K', '169K', '16.4K', '25K', '29K', '1.3K', '75K', '13.2K', '97K', '5.4K', '13.7K', '22K', '13.7K', '21K', '23K', '15.5K', '60K', '935', '13.3K', '6.7K', '9.7K', '1.5K', '10.2K', '15.6K', '25K', '22K', '77K', '3.1K', '36K', '35K', '77K', '22K', '16.9K', '24K', '11.8K', '30K', '39K', '7.6K', '6.4K', '14.4K', '14.2K', '15.3K', '37K', '28K', '26K', '5.6K']
['65K', '6.6K', '15K', '9.3K', '9.6K', '25K', '7.8K', '31K', '14.2K', '9.1K', '11.2K', '7.5K', '12.6K', '6.5K', '16.3K', '13.1K', '12.9K', '10K', '6.2K', '4.4K', '15.8K', '2.2K', '4.2K', '2.7K', '4.8K', '2.1K', '3.2K', '7.5K', '5.9K', '9.2K', '17.8K', '3.1K', '7.3K', '11.3K', '15.6K', '14.6K', '3.7K', '8.3K', '6.4K', '3K', '19.2K', '3.6K', '2.6K', '11.5K', '10.9K', '5.3K', '6.2K', '11.7K', '8.1K', '3.1K']
['8.9K', '1.2K', '6.9K', '78K', '7K', '139K', nan, nan, '2.3K', nan, '83K', '1.4K', '5.2K', '5.2K', nan, nan, '7.2K', '300', nan, '1.8K', '14.5K', nan, '300', nan, '3K', nan, nan, nan, '58K', '1.7K', 

In [67]:
print(len(names), len(exp), len(athlete_grade), len(ages), len(ranks), len(high_school), len(home_town), len(colleges), len(college_status), len(college_distance), len(num_offers), len(NIL_val),len(instagram_followers),
      len(twitter_followers),len(tiktok_followers))
#print(names[253], exp[-2], athlete_grade[-2], ages[-2], ranks[-2], high_school[-2], home_town[-2])

50 50 50 50 50 50 50 50 50 50 50 50 50 50 50


In [17]:
for athlete_link in athlete_links[250:260]:
    print(athlete_link)

https://www.on3.com/db/bobby-durkin-156822/
https://www.on3.com/db/john-blackwell-153825/
https://www.on3.com/db/comeh-emuobor-147246/
https://www.on3.com/db/malik-olafioye-81952/
https://www.on3.com/db/wesley-tubbs-iii-150899/
https://www.on3.com/db/jalen-hooks-54341/
https://www.on3.com/db/austin-ball-147266/
https://www.on3.com/db/chuck-bailey-iii-147243/
https://www.on3.com/db/gabe-sisk-44602/
https://www.on3.com/db/isaiah-manning-147244/


In [12]:
# sports = ['basketball']
# #years = ['2022','2023','2024','2025','2026']
# years = ['2023']


# urls = {}

# for year in years:
#     for sport in sports:

        
            
        # page = requests.get(URL)
        # soup = BeautifulSoup(page.content, 'html.parser')
        #results = soup.find(class_="PlayerRankings_playerRankings__jvfFg")
        
        # page_soup = soup(driver.page_source, 'html.parser')
        # print(page_soup)
        # results = page_soup.find(class_="PlayerRankings_playerRankings__jvfFg")
        # try:
        #     names_sport = results.find_all('a', class_="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-h5 MuiTypography-colorPrimary")
        #     for name in names_sport:
        #         urls[name.text] = name['href']
        # except:
        #     print("Whoops! No athletes appear to have NIL deal for {} {}".format(level, sport))
        #     continue
    #end
#end

In [13]:
# all_info = {}

# for name, url in urls.items():
#     all_info[name] = []
#     URL = f'https://www.on3.com{url}/recruiting'
#     page = requests.get(URL)
#     soup = BeautifulSoup(page.content, 'html.parser')
    
    
    
#     results = soup.find(class_="PlayerRecruiting_playergrid__wYlb_")
#     player_info = results.find_all('span', class_="MuiTypography-root MeasurementInfo_text__dCryI MuiTypography-body1 MuiTypography-colorTextPrimary")
#     for info in player_info:
#         all_info[name].append(info.text)
    
#     results = soup.find(class_="Rankings_container__U2afk")
#     rankings_info = results.find_all('p', class_="MuiTypography-root Rankings_industryRating__9uavm MuiTypography-body1 MuiTypography-colorTextPrimary")
#     for info in rankings_info:
#         all_info[name].append(info.text)
    
    
#     results = soup.find(class_="PlayerInterestsModule_targets__rxz4U")
#     recruitment_info = results.find_all(['a','h6'], class_=["MuiTypography-root MuiLink-root MuiLink-underlineNone PlayerInterestsItem_teamName__FeBHv MuiTypography-h5 MuiTypography-colorPrimary","MuiTypography-root PlayerInterestsItem_predictionLabel__mT201 MuiTypography-subtitle1 MuiTypography-colorTextPrimary"])
#     for info in  recruitment_info:
#         all_info[name].append(info.text)
